tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,434 @@
+            try
             {
-                // We found a trace parent (we are reading from the Http Headers)
+                sb = StringBuilderCache.Acquire(100);
+                sb.Append("dd=");
+
+                if (samplingPriority != null)
+                {
+                    sb.Append("s:").Append(samplingPriority).Append(';');
+                }
 
-                /* (https://www.w3.org/TR/trace-context/)
-                    Valid traceparent when caller sampled this request:
+                if (!string.IsNullOrWhiteSpace(context.Origin))
+                {
+                    sb.Append("o:").Append(context.Origin).Append(';');
+                }
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 01  // sampled
-                    Valid traceparent when caller didn’t sample this request:
+                if (context.TraceContext?.Tags?.ToArray() is { Length: > 0 } tags)
+                {
+                    foreach (var tag in tags)
+                    {
+                        if (tag.Key.StartsWith("_dd.p.", StringComparison.Ordinal))
+                        {
+#if NETCOREAPP
+                            var key = tag.Key.AsSpan(start: 6);
+#else
+                            var key = tag.Key.Substring(startIndex: 6);
+#endif
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 00  // not sampled
-                 */
+                            sb.Append("t.").Append(key).Append(':').Append(tag.Value).Append(';');
+                        }
+                    }
+                }
 
-                if (traceParent!.Length != 55 || traceParent[2] != '-' || traceParent[35] != '-' || traceParent[52] != '-')
+                if (sb.Length == 3)
                 {
-                    return false;
+                    // "dd=", we never appended anything
+                    return string.Empty;
                 }
 
-                char w3cSampled = traceParent[54];
-                if (traceParent[53] != '0' || (w3cSampled != '0' && w3cSampled != '1'))
+                // remove trailing ";"
+                if (sb[sb.Length - 1] == ';')
                 {
-                    return false;
+                    sb.Length--;
                 }
 
-                var samplingPriority = w3cSampled == '0' ? 0 : 1;
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+            finally
+            {
+                StringBuilderCache.Release(sb);
+            }
```

Comment:
I'm not sure, is it safe to call `Release` twice on the same `StringBuilder` 🤔

Reply:
Good catch! I think you are right. The first `StringBuilderCache.GetStringAndRelease(sb)` saves the reference into the thread static, making it available to the next caller of `Acquire()`. If someone acquires that reference and then we call `StringBuilderCache.Release(sb)`, we will be making it available again, although it may not be.

IIRC, a few months ago I saw someone argue in favor of using `finally` to ensure the SB is placed back in the cache, but that must have been using something like this, right?
```
    // not GetStringAndRelease(sb)
    return sb.ToString(); 
}
finally
{
    StringBuilderCache.Release(sb);
}
```

Reply:
> The first `StringBuilderCache.GetStringAndRelease(sb)` saves the reference into the thread static, making it available to the next caller of `Acquire()`. If someone acquires that reference and then we call `StringBuilderCache.Release(sb)`, we will be making it available again, although it may not be.

On second thought, I don't think there is a race condition because the field is thread static. That's the whole point: we keep it thread safe without requiring any locking mechanism, as the expense of keeping as many SB instances alive as there are threads.

Regardless, I change this to use the code above (`ToString()` and `Release()`).